File: src/ttft_itl_benchmark/metrics.py
from __future__ import annotations

from dataclasses import asdict, dataclass
import math
import statistics

from .client import RequestTrace


def percentile(values: list[float], q: float) -> float | None:
    if not values:
        return None
    if not 0 <= q <= 100:
        raise ValueError("percentile must be in [0, 100]")
    ordered = sorted(values)
    if len(ordered) == 1:
        return ordered[0]
    rank = (len(ordered) - 1) * q / 100
    low = math.floor(rank)
    high = math.ceil(rank)
    if low == high:
        return ordered[low]
    fraction = rank - low
    return ordered[low] * (1 - fraction) + ordered[high] * fraction


@dataclass(frozen=True)
class LevelSummary:
    concurrency: int
    attempted: int
    succeeded: int
    failed: int
    duration_s: float
    request_throughput_rps: float
    output_throughput_tps: float | None
    total_throughput_tps: float | None
    mean_prompt_tokens: float | None
    mean_output_tokens: float | None
    ttft_p50_ms: float | None
    ttft_p90_ms: float | None
    ttft_p95_ms: float | None
    ttft_p99_ms: float | None
    itl_p50_ms: float | None
    itl_p90_ms: float | None
    itl_p95_ms: float | None
    itl_p99_ms: float | None
    tpot_p50_ms: float | None
    tpot_p99_ms: float | None
    e2e_p50_ms: float | None
    e2e_p99_ms: float | None

    def as_dict(self) -> dict:
        return asdict(self)


def _ms(values: list[float], q: float) -> float | None:
    value = percentile(values, q)
    return None if value is None else value * 1000


def summarize_level(
    *, concurrency: int, traces: list[RequestTrace], duration_s: float
) -> LevelSummary:
    successful = [trace for trace in traces if trace.success]
    ttft = [trace.ttft_s for trace in successful if trace.ttft_s is not None]
    itl = [sample for trace in successful for sample in trace.inter_chunk_s]
    tpot = [trace.tpot_s for trace in successful if trace.tpot_s is not None]
    e2e = [trace.e2e_s for trace in successful]
    prompt_counts = [
        trace.prompt_tokens
        for trace in successful
        if trace.prompt_tokens is not None
    ]
    output_counts = [
        trace.output_tokens
        for trace in successful
        if trace.output_tokens is not None
    ]

    output_tps = None
    total_tps = None
    if duration_s > 0 and len(output_counts) == len(successful) and successful:
        output_tps = sum(output_counts) / duration_s
        if len(prompt_counts) == len(successful):
            total_tps = (sum(prompt_counts) + sum(output_counts)) / duration_s

    return LevelSummary(
        concurrency=concurrency,
        attempted=len(traces),
        succeeded=len(successful),
        failed=len(traces) - len(successful),
        duration_s=duration_s,
        request_throughput_rps=0.0 if duration_s <= 0 else len(successful) / duration_s,
        output_throughput_tps=output_tps,
        total_throughput_tps=total_tps,
        mean_prompt_tokens=statistics.fmean(prompt_counts) if prompt_counts else None,
        mean_output_tokens=statistics.fmean(output_counts) if output_counts else None,
        ttft_p50_ms=_ms(ttft, 50),
        ttft_p90_ms=_ms(ttft, 90),
        ttft_p95_ms=_ms(ttft, 95),
        ttft_p99_ms=_ms(ttft, 99),
        itl_p50_ms=_ms(itl, 50),
        itl_p90_ms=_ms(itl, 90),
        itl_p95_ms=_ms(itl, 95),
        itl_p99_ms=_ms(itl, 99),
        tpot_p50_ms=_ms(tpot, 50),
        tpot_p99_ms=_ms(tpot, 99),
        e2e_p50_ms=_ms(e2e, 50),
        e2e_p99_ms=_ms(e2e, 99),
    )

File: internal/bench/metrics.go
package bench

import (
	"math"
	"sort"
)

func Percentile(values []float64, q float64) *float64 {
	if len(values) == 0 {
		return nil
	}
	if q < 0 || q > 100 {
		panic("percentile must be in [0,100]")
	}
	ordered := append([]float64(nil), values...)
	sort.Float64s(ordered)
	if len(ordered) == 1 {
		v := ordered[0]
		return &v
	}
	rank := float64(len(ordered)-1) * q / 100
	lo := int(math.Floor(rank))
	hi := int(math.Ceil(rank))
	v := ordered[lo]
	if lo != hi {
		f := rank - float64(lo)
		v = ordered[lo]*(1-f) + ordered[hi]*f
	}
	return &v
}

func ms(values []float64, q float64) *float64 {
	p := Percentile(values, q)
	if p == nil {
		return nil
	}
	v := *p * 1000
	return &v
}

func Summarize(concurrency int, traces []Trace, durationSeconds float64) LevelSummary {
	var ttft, itl, tpot, e2e []float64
	success := 0
	allPromptCounts := true
	allOutputCounts := true
	promptTotal := 0
	outputTotal := 0

	for _, t := range traces {
		if !t.Success {
			continue
		}
		success++
		e2e = append(e2e, t.E2ESeconds)
		if t.TTFTSeconds != nil {
			ttft = append(ttft, *t.TTFTSeconds)
		}
		itl = append(itl, t.ITLSeconds...)
		if v := t.TPOTSeconds(); v != nil {
			tpot = append(tpot, *v)
		}
		if t.PromptTokens == nil {
			allPromptCounts = false
		} else {
			promptTotal += *t.PromptTokens
		}
		if t.OutputTokens == nil {
			allOutputCounts = false
		} else {
			outputTotal += *t.OutputTokens
		}
	}

	rps := 0.0
	if durationSeconds > 0 {
		rps = float64(success) / durationSeconds
	}
	var outputTPS, totalTPS *float64
	if success > 0 && durationSeconds > 0 && allOutputCounts {
		v := float64(outputTotal) / durationSeconds
		outputTPS = &v
		if allPromptCounts {
			t := float64(promptTotal+outputTotal) / durationSeconds
			totalTPS = &t
		}
	}

	return LevelSummary{
		Concurrency:          concurrency,
		Attempted:            len(traces),
		Succeeded:            success,
		Failed:               len(traces) - success,
		DurationSeconds:      durationSeconds,
		RequestThroughputRPS: rps,
		OutputThroughputTPS:  outputTPS,
		TotalThroughputTPS:   totalTPS,
		TTFTP50MS:            ms(ttft, 50),
		TTFTP90MS:            ms(ttft, 90),
		TTFTP95MS:            ms(ttft, 95),
		TTFTP99MS:            ms(ttft, 99),
		ITLP50MS:             ms(itl, 50),
		ITLP90MS:             ms(itl, 90),
		ITLP95MS:             ms(itl, 95),
		ITLP99MS:             ms(itl, 99),
		TPOTP50MS:            ms(tpot, 50),
		TPOTP99MS:            ms(tpot, 99),
		E2EP50MS:             ms(e2e, 50),
		E2EP99MS:             ms(e2e, 99),
	}
}
